File: clonewhatapp/CloneWhatsApp.API/Models/Mensaje.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace CloneWhatsApp.API.Models;

public enum TipoMensaje
{
    Texto,
    Imagen,
    Video,
    Audio,
    Documento,
    Ubicacion,
    Contacto,
    Sistema
}

public enum EstadoMensaje
{
    Enviado,
    Entregado,
    Leido
}

public class Mensaje
{
    [Key]
    public Guid Id { get; set; }

    [Required]
    public Guid ChatId { get; set; }

    [Required]
    public Guid RemitenteId { get; set; }

    [MaxLength(4000)]
    public string? Contenido { get; set; }

    public TipoMensaje Tipo { get; set; } = TipoMensaje.Texto;

    [MaxLength(500)]
    public string? UrlArchivo { get; set; }

    [MaxLength(100)]
    public string? NombreArchivo { get; set; }

    public long? TamanoArchivo { get; set; }

    // Para mensajes de audio
    public int? DuracionSegundos { get; set; }

    // Para respuestas
    public Guid? MensajeRespondidoId { get; set; }

    public DateTime FechaEnvio { get; set; } = DateTime.UtcNow;

    public bool Eliminado { get; set; } = false;

    public bool EliminadoParaTodos { get; set; } = false;

    public bool Editado { get; set; } = false;

    // Navigation properties
    [ForeignKey("ChatId")]
    public Chat Chat { get; set; } = null!;

    [ForeignKey("RemitenteId")]
    public Usuario Remitente { get; set; } = null!;

    [ForeignKey("MensajeRespondidoId")]
    public Mensaje? MensajeRespondido { get; set; }

    public ICollection<EstadoMensajeUsuario> Estados { get; set; } = new List<EstadoMensajeUsuario>();
}

File: clonewhatapp/whatsapp-clone-android/domain/src/main/java/com/clonewhatsapp/domain/model/Mensaje.kt
package com.clonewhatsapp.domain.model

data class Mensaje(
    val id: String,
    val chatId: String,
    val remitenteId: String,
    val contenido: String? = null,
    val urlArchivo: String? = null,
    val tipo: TipoMensaje = TipoMensaje.TEXTO,
    val estado: EstadoMensaje = EstadoMensaje.ENVIADO,
    val fechaEnvio: Long,
    val editado: Boolean = false,
    val eliminado: Boolean = false,
    val eliminadoParaTodos: Boolean = false,
    val mensajeRespondidoId: String? = null,
    val duracionSegundos: Int? = null,
    val nombreRemitente: String? = null,
)

enum class TipoMensaje {
    TEXTO,
    IMAGEN,
    VIDEO,
    AUDIO,
    DOCUMENTO,
    UBICACION,
    CONTACTO,
    SISTEMA,
}

enum class EstadoMensaje {
    ENVIADO,
    ENTREGADO,
    LEIDO,
}

enum class SyncStatus {
    PENDING,
    SYNCED,
    FAILED,
}
